controllers/humiocluster_controller.go

```diff
@@ -2095,18 +2119,348 @@ func (r *HumioClusterReconciler) ensurePodsExist(ctx context.Context, hc *humiov
 	return reconcile.Result{}, nil
 }
 
+func (r *HumioClusterReconciler) processDownscaling(ctx context.Context, hc *humiov1alpha1.HumioCluster, hnp *HumioNodePool, req ctrl.Request) (reconcile.Result, error) {
+	labelsToMatch := hnp.GetNodePoolLabels()
+	labelsToMatch[kubernetes.PodMarkedForDataEviction] = "false"
+	podsNotMarkedForEviction, err := kubernetes.ListPods(ctx, r, hnp.GetNamespace(), labelsToMatch)
+	if err != nil {
+		return reconcile.Result{}, r.logErrorAndReturn(err, "failed to list pods")
+	}
+
+	labelsToMatch[kubernetes.PodMarkedForDataEviction] = "true"
+	podsMarkedForEviction, err := kubernetes.ListPods(ctx, r, hnp.GetNamespace(), labelsToMatch)
+	if err != nil {
+		return reconcile.Result{}, r.logErrorAndReturn(err, "failed to list pods")
+	}
+
+	clusterConfig, err := helpers.NewCluster(ctx, r, hc.Name, "", hc.Namespace, helpers.UseCertManager(), true, false)
+	if err != nil {
+		return reconcile.Result{}, r.logErrorAndReturn(err, "could not create a cluster config for the http client.")
+	}
+	humioHttpClient := r.HumioClient.GetHumioHttpClient(clusterConfig.Config(), req)
+
+	// remove lingering nodes
+	for _, vhost := range hc.Status.EvictedNodeIds {
+		err = r.unregisterNode(ctx, hc, humioHttpClient, req, vhost)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// If there are more pods than specified, evict pod
+	if len(podsNotMarkedForEviction) > hnp.GetNodeCount() && len(podsMarkedForEviction) == 0 { // mark a single pod, to slowly reduce the node count.
+		err := r.markPodForEviction(ctx, hc, req, podsNotMarkedForEviction, hnp.GetNodePoolName())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if there are pods marked for eviction
+	if len(podsMarkedForEviction) > 0 {
+		// check the eviction process
+		for _, pod := range podsMarkedForEviction {
+			vhostStr := pod.Annotations[kubernetes.LogScaleClusterVhost]
+			vhost, err := strconv.Atoi(vhostStr)
+			if err != nil {
+				return reconcile.Result{}, r.logErrorAndReturn(err, fmt.Sprintf("could not parse vhost from annotation %s", vhostStr))
+			}
+			nodeCanBeSafelyUnregistered, err := r.checkEvictionStatusForPod(ctx, humioHttpClient, req, vhost)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			if nodeCanBeSafelyUnregistered {
+				r.Log.Info(fmt.Sprintf("successfully evicted data from vhost %d", vhost))
+				hc.Status.EvictedNodeIds = append(hc.Status.EvictedNodeIds, vhost) // keep track of the evicted node for unregistering
+				err = r.Status().Update(ctx, hc)
+				if err != nil {
+					r.Log.Error(err, "failed to update cluster status")
+					return reconcile.Result{}, err
+				}
+				r.Log.Info(fmt.Sprintf("removing pod %s containing vhost %d", pod.Name, vhost))
+				if err := r.Delete(ctx, &pod); err != nil { // delete pod before unregistering node
+					return reconcile.Result{}, r.logErrorAndReturn(err, fmt.Sprintf("failed to delete pod %s for vhost %d!", pod.Name, vhost))
+				}
```

Comment:
What happens if the `r.Status().Update(ctx, hc)` call succeeds, but then the `r.Delete(ctx, &pod)` fails? Or perhaps we update status, and then due to the operator being restarted, it doesn't get to do the pod deletion?

Reply:
If I read the code correctly, it would essentially "just" mean it'll append the ID one more time to the Status, and then try to do the pod deletion again? Perhaps it should only do the append operation if the vhost is not already in the list?

Reply:
That's fine. If the delete fails after the update succeeds, then the node will remain marked at Evicted and at the next reconcile run, it will try to delete the pod again.
As per the check at [line 2228](https://github.com/humio/humio-operator/blob/30bcb8eff19572a0e364074dae4bd8896787a4a4/controllers/humiocluster_controller.go#L2228), the node will NOT be unregistered if it's still alive (if the pod is not deleted).